avalanche/benchmarks/utils/transform_groups.py

```diff
@@ -0,0 +1,128 @@
+"""
+    Transformation groups manage transformations that are in different
+    phases of the optimization process, such as different train/eval transformations.
+
+    They support multi-argument transforms as defined in
+    `avalanche.benchmarks.utils.transforms`.
+"""
+from collections import defaultdict
+from typing import Dict, Union, Callable, Sequence
+
+from avalanche.benchmarks.utils.transforms import \
+    MultiParamCompose, TupleTransform, MultiParamTranform
+
+
+class TransformGroups:
+    """Transformation groups for Avalanche datasets.
+
+    TransformGroups supports preprocessing and augmentation pipelines for
+    Avalanche datasets. Transfomations are separated into groups (e.g. `train`
+    transforms and `test` transforms), that can be easily switched using the
+    `with_transform` method.
+    """
+
+    def __init__(self,
+                 transform_groups: Dict[str, Union[Callable, Sequence[Callable]]],
+                 current_group="train"):
+        for group, transform in transform_groups.items():
+            transform = _normalize_transform(transform)
+            transform_groups[group] = transform
+        self.transform_groups = transform_groups
+        self.current_group = current_group
+
+        if "train" in transform_groups:
+            if "eval" not in transform_groups:
+                transform_groups["eval"] = transform_groups["train"]
+
+        if "train" not in transform_groups:
+            transform_groups["train"] = None
+
+        if "eval" not in transform_groups:
+            transform_groups["eval"] = None
+
+    def __getitem__(self, item):
+        return self.transform_groups[item]
+
+    def __setitem__(self, key, value):
```

Comment:
In general I don't love the idea of having mutable objects in the dataset, same for the with_transform method. The issue is that we may end up having datasets in strategy fields, plugins (especially the replay plugin), etc. that change behaviour in time.

Reply:
notice that right now the `__getitem__` pass the current group explicitly to the datasets in the hierarchy. So, the current transformation group can be safely changed without problems. However, I agree that it may be better to make transformation groups as an immutable object. We could even remove `self.current_group` from transformation groups.